File: jianzhioffer/firstround/jianzhi_53_2.py
class Solution(object):
    def missingNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        i, j = 0, len(nums)-1
        while i <= j:
            mid = (i + j) // 2
            if nums[mid] <= mid:
                i = mid + 1
            else:
                j = mid - 1
        return i


if __name__ == '__main__':
    arr = [0,1,2,4,5,6,7,8,9]
    obj = Solution()
    num = obj.missingNumber(arr)
    print(num)

File: jianzhioffer/secondround/jianzhi_53_2.cpp
#include <vector>
using namespace std;

class Solution
{
public:
    Solution();
    ~Solution();
    int missingNumber(vector<int>& nums);
};

Solution::Solution() {}

Solution::~Solution() {}

//折半查找，时间复杂度O(logN)，空间复杂度O(1)
int Solution::missingNumber(vector<int>& nums)
{
    if(nums.empty()) return -1;
    int left=0, right=nums.size()-1, mid=0;
    while (left<=right)
    {
        mid = (left+right)/2;
        if(nums[mid]==mid) left=mid+1;
        else right=mid-1;
    }
    return left;
}

int main(int argc, char const *argv[])
{
    vector<int> nums = {0};
    Solution obj;
    int result = obj.missingNumber(nums);
    return 0;
}
